### PR TITLE
[SPARK-22799][ML] Bucketizer should throw exception if single- and multi-column params are both set

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Currently there is a mixed situation when both single- and multi-column are supported. In some cases exceptions are thrown, in others only a warning log is emitted. In this discussion https://issues.apache.org/jira/browse/SPARK-8418?focusedCommentId=16275049&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16275049, the decision was to throw an exception.

The PR throws an exception in `Bucketizer`, instead of logging a warning.

## How was this patch tested?

modified UT

### CODE REVIEW

**Comment by mgaido91** [comment]
Jenkins, retest this please

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Here it is better to add one more check:
if single column, only set `splits` param.
if multiple column, only set `splitsArray` param.

**Review by hhbyyh — COMMENTED** [review]
@mgaido91, Thanks for the PR. 

Since we have several classes that need the improvement and more classes will support HasInputCols, I feel like we should develop some code that can be shared by other classes. 
We also need common unit test methods for the case.

**Comment by mgaido91** [comment]
@hhbyyh thanks for the review. I see that for some classes there are ongoing PRs. Thus I cannot change them now in order to have a common place and a common test. Should I wait for those PRs to be merged then? Or were you suggesting something different?
Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: thanks. I will add these checks while doing the changes requested by @hhbyyh. Thanks.

**Comment by hhbyyh** [comment]
I would suggest to develop the common infrastructure and unit test first, then other PR can take it or we can send follow-up fix.

cc @MLnick for advice.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: We should modify the document above too.

**Comment by mgaido91** [comment]
@hhbyyh I created a common infrastructure. Please let me know if I have to modify it. Meanwhile, I'd like to discuss where to put the common UTs: do you have any specific idea about the right place? Thanks.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I noticed `isBucketizeMultipleColumns` is invoked in many places and maybe we can put the checks in other places like transformSchema. It also makes the code consistent with function name.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: This may not necessarily be an error for some classes, but we can keep it for now.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: private[spark]

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: private[spark]

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Error message can be more straight forward. e.g. `$paramName1` and `$paramName2` cannot be set simultaneously.

**Review by hhbyyh — COMMENTED** [review]
Thanks for the update. 

I like the current implementation that can be shared by other classes. For the test case, I think it makes sense to add something in `checkParams`.

cc @MLnick @jkbradley @srowen

**Comment by mgaido91** [comment]
@hhbyyh thanks for the comments. I already fixed all your comments, but I am waiting to push for the UT. Honestly I think that `checkParam` is not the best place. Checking the exception requires setting the parameters and invoking `transform`, thus I am not sure it is the best place, since it is a very generic one and at the moment we are setting no parameter there. What do you think?

**Comment by hhbyyh** [comment]
To make it available for other classes, we need to support checking for both `fit` and `transform`, that means we also need a sample input Dataset, so we may have to add the explicit test in each of the test suite. But we can still create some infrastructure function for the explicit test to invoke.
E.g. we can create some function in object ParamsSuite or other places 
`checkMultiColumnParams(obj: Params, sampleData: Dataset[_])`

**Comment by mgaido91** [comment]
thanks @hhbyyh, I updated the PR according to your suggestion and previous comments.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: suggestion for function name:
assertColOrCols --> checkMultiColumnParams

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala
Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: suggestion for function name:
assertColOrCols --> testMultiColumnParams

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala
Comment (hhbyyh) [inline mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: I don't think the DataFrame here can be used for other transformers like StringIndexer.

How about add a Dataset as function parameter? And is it possible to use an instance `obj: Params` rather than `paramsClass: Class[_ <: Params]` as parameter, just to be more flexible.

**Review by hhbyyh — COMMENTED** [review]
Thanks. Add some suggestions for function names and test utility.

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala
Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: The reason why I created the dataframe inside the method was to control the names of the columns it has. Otherwise we can't ensure that those columns exist. I think that the type check is performed later, thus it is  not a problem here. What do you think?

I preferred to use `paramsClass: Class[_ <: Params]` because I need a clean instance for each of the two checks: if an instance is passed I cannot enforce that it is clean, ie. some parameters weren't already set and I would need to copy it to create new instances as well, since otherwise the second check would be influenced by the first one. What do you think?

Thanks.

Comment (hhbyyh) [reply mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: Send column names as parameter if necessary. We need to ensure the test utility can be used by most transformers with multiple column support.

> I think that the type check is performed later, thus it is not a problem here. 

I don't quite get it, in either `transform` or `fit` the data type will be checked and they will trigger exceptions.

> I preferred to use paramsClass: Class[_ <: Params]

I'm only thinking about the case that default constructor is not sufficient to create a working Estimator/Transformer. If that's not a concern, then reflection is OK.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Sorry to miss it, but I just found that FeatureHasher has both InputCols and OutputCol. 
I think we can remove the case and the one above since they can be too strict.

**Review by hhbyyh — COMMENTED** [review]
LGTM besides the one comment.

**Comment by MLnick** [comment]
Yes featurehasher doesn’t need inputCol and inputCols - since it is a new
multi column transformer. We may see this more in future - as I think new
transformers should be able to work on multiple columns if possible

**Comment by mgaido91** [comment]
thanks @hhbyyh, I removed the 2 too strict checks.

**Comment by mgaido91** [comment]
@MLnick @viirya  @WeichenXu123 any further comments on this?

**Comment by mgaido91** [comment]
@jkbradley maybe you can also help reviewing this, thanks.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: Seems superfluous to how have a separate method for this

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: "cannot be both" -> "cannot both be"

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: "is thrown is both" -> "is thrown if both"

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: only whether" -> "only when"

**Review by MLnick — COMMENTED** [review]
A few minor comments. I'm not totally certain about the generic test case setup. I will think about it a bit and revert with further comments.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: If we need to check other exclusive params, e.g., `inputCol` and `splitsArray` or `inputCols` and `splits`, why not just have a method like:

```scala
def checkExclusiveParams(model: Params, params: String*): Unit = {
  if (params.filter(model.isSet(_)).size > 1) {
    val paramString = params.mkString("`", "`, `", "`")
    throw new IllegalArgumentException(s"$paramString are exclusive, but more than one among them are set.")
  }
}
ParamValidators.checkExclusiveParams(this, "inputCol", "inputCols")
ParamValidators.checkExclusiveParams(this, "outputCol", "outputCols")
ParamValidators.checkExclusiveParams(this, "inputCol", "splitsArray")
ParamValidators.checkExclusiveParams(this, "inputCols", "splits")
```

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: I added this method too in #20146.

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: I think we can use that method once merged, thanks.

Comment (MLnick) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: I am not sure if #20146 will get merged for `2.3` - but I think we must merge this PR for `2.3` because I'd prefer not to have this inconsistency in param error handling between `QuantileDiscretizer` and `Bucketizer`. This is a relatively small change, so we can merge it into the branch if we move it quickly.

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: We should also test the other exclusive params (input cols and splits params) as per https://github.com/apache/spark/pull/19993/files#r159133936

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Based on https://github.com/apache/spark/pull/20146#issuecomment-357068054 from @WeichenXu123, I think #20146 cannot get merged for 2.3.

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: If this method looks good to you, maybe you can just copy it from #20146 to use here.

Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: @MLnick @viirya in order to address https://github.com/apache/spark/pull/19993/files#r161682506, I was thinking to let this method as it is (just renaming it as per @viirya suggestion) and only adding an `additionalExclusiveParams: (String, String)*` argument to the function. WDYT?

Comment (MLnick) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: I think @viirya's method is simpler and more general, so why not use it?

Comment (MLnick) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: @viirya your actual method in #20146 is slightly different (see [here](https://github.com/apache/spark/pull/20146/files#diff-72f95a0938e5a140d5126f06bdc381a6R254)). Is that the best version to use?

Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: @MLnick Yes. I didn't test the method posted here. The model possibly doesn't have the params, so we need to check it with `model.hasParam`. Please use the method in #20146.

**Comment by jkbradley** [comment]
Since RC1 for 2.3 failed, it'd be great to get this into 2.3.  @mgaido91 do you mind if I send my comments along with a PR to update this PR of yours?  I'm rushing because of the time pressure to get this into 2.3 (to avoid a behavior change between 2.3 and 2.4).  Thanks in advance!

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: The problem with trying to use a general method like this is that it's hard to capture model-specific requirements.  This currently misses checking to make sure that exactly one (not just <= 1) of each pair is available, plus that all of the single-column OR all of the multi-column Params are available.  (The same issue occurs in https://github.com/apache/spark/pull/20146 )  It will also be hard to check these items and account for defaults.

I'd argue that it's not worth trying to use generic checking functions here.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Let's switch this to use the Logging trait, to match other MLlib patterns.

**Review by jkbradley — COMMENTED** [review]
I'm just leaving early comments to note the main issue I see, especially since they are relevant to https://github.com/apache/spark/pull/20146 .  I'll send a PR soon, later today.

**Comment by mgaido91** [comment]
@jkbradley sure no problem, let me know how I can help.

File: mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: my initial implementation (with @hhbyyh's comments) was more generic and checked what you said. After, @MLnick and @viirya asked to switch to a more generic approach which is the current you see. I'm fine with either of those, but I think we need to choose one way and go in that direction, otherwise we just loose time.

Comment (jkbradley) [reply mllib/src/main/scala/org/apache/spark/ml/feature/Bucketizer.scala]: I see.  I'll see if I can come up with something which is generic but handles these other checks.

**Comment by jkbradley** [comment]
OK, sent: https://github.com/mgaido91/spark/pull/1

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: Only comment I have is that I believe this line is not testing what you may think.

As I read the [checkSingleVsMultiColumnParams](https://github.com/apache/spark/pull/19993/files#diff-72f95a0938e5a140d5126f06bdc381a6R266) method, in this test case it will throw the error, _not_ because both `splits` and `splitsArray` are set, but rather because both `inputCol` & `inputCols` are _unset_.

Actually it applies to the line above too.

@jkbradley

**Comment by MLnick** [comment]
Overall looks good with @jkbradley's changes. I just left a comment on the param test cases as I think they're not quite complete

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala
Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: @MLnick actually it will fail for both reasons. We can add more test cases to check each of these two cases if you think it is needed.

**Comment by MLnick** [comment]
Well yes it would - but the method checks inputCols/inputCol first so will
always fail for that reason here, ie we aren’t actually testing the full
code path

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: I don't think these are used any longer?

File: mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala
Comment (MLnick) [inline mllib/src/test/scala/org/apache/spark/ml/feature/BucketizerSuite.scala]: Technically here we should probably also test the `inputCols` + `outputCols` case (i.e. that not setting `splitsArray` also throws an exception).

**Review by MLnick — APPROVED** [review]
A couple minor comments - but that can be left for a clean up in a follow up PR if need be.

I'd prefer to merge this to branch-2.3.

**Comment by WeichenXu123** [comment]
+1 merge this to 2.3

**Review by MLnick — APPROVED** [review]
LGTM

**Review by WeichenXu123 — APPROVED** [review]

**Comment by MLnick** [comment]
Merged to master / branch-2.3

**Comment by MLnick** [comment]
Thanks @mgaido91 and @jkbradley for working on this and others for review

**Pull request closed without merging**
